etc/sync-function-validation-module.js

```diff
@@ -475,71 +410,155 @@ function() {
+        // can't guarantee that the attachment is present when the attachment reference property is created/updated for it, so only
+        // validate it if it's present. The good news is that, because adding an attachment is a two part operation (create/update the
+        // document and add the attachment), the sync function will be run once for each part, thus ensuring the content is verified once
+        // both parts have been synced.
+        if (doc._attachments && doc._attachments[itemValue]) {
+          var attachment = doc._attachments[itemValue];
 
-        if (validator.supportedContentTypes && validator.supportedContentTypes.indexOf(attachment.content_type) < 0) {
-            validationErrors.push('attachment reference "' + buildItemPath(itemStack) + '" must have a supported content type (' + validator.supportedContentTypes.join(',') + ')');
+          if (validator.supportedContentTypes && validator.supportedContentTypes.indexOf(attachment.content_type) < 0) {
+              validationErrors.push('attachment reference "' + buildItemPath(itemStack) + '" must have a supported content type (' + validator.supportedContentTypes.join(',') + ')');
+          }
+
+          if (!isValueNullOrUndefined(validator.maximumSize) && attachment.length > validator.maximumSize) {
+            validationErrors.push('attachment reference "' + buildItemPath(itemStack) + '" must not be larger than ' + validator.maximumSize + ' bytes');
+          }
         }
+      }
+    }
 
-        if (!isValueNullOrUndefined(validator.maximumSize) && attachment.length > validator.maximumSize) {
-          validationErrors.push('attachment reference "' + buildItemPath(itemStack) + '" must not be larger than ' + validator.maximumSize + ' bytes');
+    function validateAttachments() {
+      var maximumIndividualAttachmentSize =
+        docDefinition.attachmentConstraints ? docDefinition.attachmentConstraints.maximumIndividualSize : null;
+      var maximumTotalAttachmentSize = docDefinition.attachmentConstraints ? docDefinition.attachmentConstraints.maximumTotalSize : null;
+      var maximumAttachmentCount = docDefinition.attachmentConstraints ? docDefinition.attachmentConstraints.maximumAttachmentCount : null;
+
+      var totalSize = 0;
+      var attachmentCount = 0;
+      for (var attachmentName in doc._attachments) {
+        attachmentCount++;
+
+        var attachmentSize = doc._attachments[attachmentName].length;
+        totalSize += attachmentSize;
+
+        if (isInteger(maximumIndividualAttachmentSize) && attachmentSize > maximumIndividualAttachmentSize) {
+          // If this attachment is owned by an attachment reference property, that property's size constraint (if any) takes precedence
+          var attachmentRefValidator = attachmentReferenceValidators[attachmentName];
+          if (isValueNullOrUndefined(attachmentRefValidator) || !isInteger(attachmentRefValidator.maximumSize)) {
```

Comment:
I'm assuming the `else` case here is instead handled by the attachment reference property validator in it's respective function?